lxd/patches.go

```diff
@@ -1420,4 +1421,34 @@ func patchInstanceRemoveVolatileLastStateIPAddresses(_ string, d *Daemon) error
 	return nil
 }
 
+func patchRemoveEntityTypeCertificate(_ string, d *Daemon) error {
+	s := d.State()
+	err := s.DB.Cluster.Transaction(s.ShutdownCtx, func(ctx context.Context, tx *db.ClusterTx) error {
+		// Using "UPDATE OR REPLACE" in case the uniqueness constraint on auth_groups_permissions is violated by the update.
+		// This would be because two permissions were granted on a group that reference the same identity. E.g.:
+		// 	lxc auth group permission add <group> certificate <fingerprint> can_view
+		//	lxc auth group permission add <group> identity tls/<fingerprint> can_view
+		// Even though these reference the same row in the identities table, the permissions are different because the
+		// entity type codes are different. Updating the code of the certificate permission would then cause a uniqueness
+		// constraint violation.
+		_, err := tx.Tx().Exec(`UPDATE OR REPLACE auth_groups_permissions SET entity_type = 24 WHERE entity_type = 4`)
+		if err != nil {
+			return fmt.Errorf("Failed replacing certificate entities with identity entities in permissions table: %w", err)
+		}
+
+		// Not using "OR REPLACE" here because warnings do not have a uniqueness constraint applied to the entity_type_code column.
+		_, err = tx.Tx().Exec(`UPDATE warnings SET entity_type_code = 24 WHERE entity_type_code = 4`)
```

Comment:
So this could leave two identical warnings?

Reply:
I'm not aware of any warnings being introduced for the `identity` entity type. So any existing warnings for certificates will be moved there.

Reply:
Oh cool then, thanks for the info